### PR TITLE
chunked: refactor file system ops in a new file and add some tests

### PR DESCRIPTION
**Pull request by giuseppe**
more details in each commit

### CODE REVIEW

**Comment by giuseppe** [comment]
@cgwalters @rhatdan PTAL

File: pkg/chunked/filesystem_linux.go [hunk 1]
```diff
@@ -304,7 +304,7 @@ func openFileUnderRoot(dirfd int, name string, flags uint64, mode os.FileMode) (
 	if errors.Is(err, unix.ENOENT) && hasCreate {
 		parent := filepath.Dir(name)
 		if parent != "" {
-			newDirfd, err2 := openOrCreateDirUnderRoot(parent, dirfd, 0)
+			newDirfd, err2 := openOrCreateDirUnderRoot(dirfd, parent, 0)
```
Comment (cgwalters) [inline pkg/chunked/filesystem_linux.go:307]: On a related topic, wouldn't it make sense to create a struct wrapper for the dirfd, and have things be methods on it?

In my Rust code, we *extensively* use https://docs.rs/cap-std/latest/cap_std/ which is related, and has a `Dir` struct with a lot of methods, but in implementation it's just:

https://github.com/bytecodealliance/cap-std/blob/1dc17d327bdcd44a0f5ec53d9fc671285be55b34/cap-std/src/fs/dir.rs#L48-L50

File: pkg/chunked/filesystem_linux_test.go [hunk 2]
```diff
@@ -50,3 +56,305 @@ func TestSeekableFileGetBlobAt(t *testing.T) {
...
+	tmpDir := createTempDir(t)
+	defer os.RemoveAll(tmpDir)
+
+	srcFile := createTempFile(t, tmpDir, "source")
+	defer srcFile.Close()
+	srcFd := int(srcFile.Fd())
+
+	destDir := createTempDir(t)
+	defer os.RemoveAll(destDir)
+	destDirFd, err := syscall.Open(destDir, syscall.O_RDONLY, 0)
```
Comment (cgwalters) [inline pkg/chunked/filesystem_linux_test.go:70]: Hello, your friendly `O_CLOEXEC` linter reporting in here :smile:

File: pkg/chunked/filesystem_linux.go [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+	doChmod := func() error {
+		if usePath {
+			return unix.Fchmodat(dirfd, baseName, uint32(mode), unix.AT_SYMLINK_NOFOLLOW)
+		}
+		return unix.Fchmod(fd, uint32(mode))
+	}
+
+	if err := doChown(); err != nil {
+		if !options.IgnoreChownErrors {
+			return &fs.PathError{Op: "chown", Path: metadata.Name, Err: err}
```
Comment (mtrmac) [inline pkg/chunked/filesystem_linux.go:170]: … and, symmetrically to how `os.Stat` returns a `PathError`, I’d expect the `PathError` creation to happen at a fairly low level.

---

I don’t know how I feel about a `PathError` reporting a `Path` which is _not exactly what we used_. 

---

Also, the commit says “improve error messages”. but it removes information about non-path parameters to the operation. Is that intentional?

(I didn’t review more parts of the “improve error messages” commit in detail.)

File: pkg/chunked/filesystem_linux_test.go [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+	err = newFile.Close()
+	assert.NoError(t, err)
+
+	// the file was copied with hard links, the inodes must be equal
+	assert.Equal(t, st.Dev, st2.Dev)
+	assert.Equal(t, st.Ino, st2.Ino)
+}
+
+func createTempDir(t *testing.T) string {
```
Comment (mtrmac) [inline pkg/chunked/filesystem_linux_test.go:350]: Just use `t.TempDir()` in callers, and there’s no need to worry about cleanup.

File: pkg/chunked/filesystem_linux_test.go [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+	tmpDir, err := os.MkdirTemp("", "testDir")
+	assert.NoError(t, err)
+	return tmpDir
+}
+
+func createTempFile(t *testing.T, dir, name string) *os.File {
+	tmpFile, err := os.CreateTemp(dir, name)
+	assert.NoError(t, err)
```
Comment (mtrmac) [inline pkg/chunked/filesystem_linux_test.go:358]: (Test case setup should typically use `require.NoError` when there’s no point in proceeding. But that can always be fixed later, if the test ever actually fail.)

**Review by mtrmac — COMMENTED** [review]
A quick skim:

- I like the consistent (parentFD, name) parameter ordering
- More tests are always good

(I didn’t carefully read the tests, and didn’t analyze the error reporting change impact in  detail.)

File: pkg/chunked/filesystem_linux.go (hunk 1 above)
Comment (giuseppe) [reply pkg/chunked/filesystem_linux.go:307]: we need an explicit `mode` to support `ForceMask`

**Comment by giuseppe** [comment]
thanks for the comments and suggestions.

Addressed them and pushed a new version

File: pkg/chunked/filesystem_linux.go [hunk 6]
```diff
@@ -58,7 +63,7 @@ func copyFileContent(srcFd int, fileMetadata *fileMetadata, dirfd int, mode os.F
 	src := fmt.Sprintf("/proc/self/fd/%d", srcFd)
 	st, err := os.Stat(src)
 	if err != nil {
-		return nil, -1, fmt.Errorf("copy file content for %q: %w", destFile, err)
+		return nil, -1, err
```
Comment (rhatdan) [inline pkg/chunked/filesystem_linux.go:66]: Not sure this is an improvement on the error. 
Now we will get "failed to stat /proc/self/fd/%d" Rather then destFile?

Comment (giuseppe) [reply pkg/chunked/filesystem_linux.go]: I've dropped the patch from this PR.

I will rework errors later

File: pkg/chunked/filesystem_linux_test.go [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+		[]byte("World"),
+	}
+
+	i := 0
+	for stream := range streams {
+		buf := new(bytes.Buffer)
+		_, err := buf.ReadFrom(stream)
+		require.NoError(t, err)
+
+		require.Equal(t, buf.Bytes(), expectedContents[i])
```
Comment (mtrmac) [inline pkg/chunked/filesystem_linux_test.go:51]: (FWIW, order is `Equal(t, expected, actual)`. Not generally worth worrying about, that that can be cleaned up if the test ever fails.)

File: pkg/chunked/filesystem_linux.go [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		destDirFd = int(f.Fd())
+	}
+
+	err = doHardLink(int(sourceFile.Fd()), destDirFd, destBase)
+	if err != nil {
+		return err
```
Comment (mtrmac) [inline pkg/chunked/filesystem_linux.go:415]: Was the removal of error wrapping here intentional?

**Review by mtrmac — COMMENTED** [review]
LGTM

File: pkg/chunked/filesystem_linux_test.go (hunk 7 above)
Comment (giuseppe) [reply pkg/chunked/filesystem_linux_test.go]: fixed now

File: pkg/chunked/filesystem_linux.go (hunk 8 above)
Comment (giuseppe) [reply pkg/chunked/filesystem_linux.go]: no, it was not.  Thanks for catching it.  I'll deal with errors in a separate PR.  I've re-added it

**Review by mtrmac — COMMENTED** [review]
/lgtm

**Pull request merged**